PR para resolver a issue #5 - Scan da imagem docker

mmmarceleza opened the pull request:
Prezados, segue o PR para o _scan_ da imagem, referente à issue #5 

O último _job_ que adicionei quebra o _CI_ porque encontra vulnerabilidades na imagem gerada no _build_. Fiz um teste local aqui com a versão seguinte (`1.26`) e não encontrei problemas do nível `HIGH` ou `CRITICAL`. Sugiro subir a versão para `1.26`.

Outra observação, há como alterar a saída do CI de `fail` para `warnig` quando encontrar uma saída de erro? Se sim, daria para colocar dessa forma ao encontrar vulnerabilidades nesse _job_, o que acham?

Também fiz uma melhoria no README, me passem o _feedback_ se ficou bom.

mmmarceleza committed:
Orgaização dos jobs e steps no CI

Adicionei algumas separações para deixar mais evidente a utilização dos jobs e steps no CI.

mmmarceleza committed:
Melhoria nas descrições do arquivo README

- Inclusão do objetivo, bem básico.
- Detalhamento dos passos do CI

mmmarceleza committed:
Adição do job para scanear a imagem docker

Utilização da ferramenta open source chamada Trivy para realizar o scan da imagem.

mmmarceleza committed:
Colocando o scan como último job

mmmarceleza committed:
correções nas ordens das numerações

mmmarceleza committed:
fix: correção no README

gomex commented on .github/workflows/main.yml:
```diff
@@ -88,7 +111,32 @@ jobs:
 
+#------------------------------------------------------
+#              job 04 - Scan da imagem
+#------------------------------------------------------
+  build:
+    name: Scan da Imagem
+    needs: push
+    runs-on: ubuntu-18.04
+    steps:
+#---------------------step 04-01-----------------------
+      - name: Checkout code
+        uses: actions/checkout@v2
+#---------------------step 04-02-----------------------
+      - name: Build an image from Dockerfile
+        run: |
+          docker build -t docker.io/my-organization/my-app:${{ github.sha }} .
+#---------------------step 04-03-----------------------
+      - name: Scanear Imagem com Trivy
+        uses: aquasecurity/trivy-action@master
+        with:
+          image-ref: 'docker.io/my-organization/my-app:${{ github.sha }}'
+          format: 'table'
+          exit-code: '1'
+          ignore-unfixed: true
+          vuln-type: 'os,library'
+          severity: 'CRITICAL,HIGH'
```
Isso pode funcionar antes do push? Se sim, seria o ideal.

mmmarceleza replied:
Pode sim, roda independentemente dos outros _jobs_ se quisermos. Eu acabei deixando no fim porque imaginei que seria usado apenas como um informativo. Pode acontecer de surgir uma vulnerabilidade que ainda não tem correção. Podemos discutir isso com calma depois.

gomex reviewed (changes requested):
Está ficando muito bom!

mmmarceleza committed:
fix: correção do PR #6

Alteração da ordem do scan da imagem e a dependência do push para o Docker Hub apenas com a aprovação da scan.

mmmarceleza committed:
adição da figura do fluxo do CI

mmmarceleza commented:
@gomex, fiz as correções. Só fiquei na dúvida se a publicação da imagem para Docker Hub era nos _commits_ para o `main` ou na criação de tags. Na semana passada eu deveria ter anotado.

edsoncelio commented:
> @gomex, fiz as correções. Só fiquei na dúvida se a publicação da imagem para Docker Hub era nos _commits_ para o `main` ou na criação de tags. Na semana passada eu deveria ter anotado.

Opa @mmmarceleza, o push para o docker hub é feito com base em branch, na `main`, nesse caso:        
 `if: github.ref == 'refs/heads/main'`

edsoncelio commented:
Muito bom @mmmarceleza, seguem alguns pontos que identifiquei/fiquei em dúvida:

1. Dado que o passo de `scan` vai executar sempre, poderíamos deixar a execução apenas em `pull_request`, com isso a gente ia garantir que `push` em `main` vai sempre ter executado esse passo (ficaria como pré-requisito) e o passo de `push` faria apenas a publicação do container, o que acha?
2. Nessas linhas ([aqui](https://github.com/mmmarceleza/iac-docker-tika/blob/babba72e1e8834415656f3db3b627e8715a72f9f/.github/workflows/main.yml#L78) e [aqui](https://github.com/mmmarceleza/iac-docker-tika/blob/babba72e1e8834415656f3db3b627e8715a72f9f/.github/workflows/main.yml#L83)) não seria melhor nomear com o nome certinho do repo? algo como `iac-docker-tika:....`
3. No seu repositório o CI está falhando, pode dar uma olhada nisso?

mmmarceleza commented:
> Muito bom @mmmarceleza, seguem alguns pontos que identifiquei/fiquei em dúvida:
> 
> 1. Dado que o passo de `scan` vai executar sempre, poderíamos deixar a execução apenas em `pull_request`, com isso a gente ia garantir que `push` em `main` vai sempre ter executado esse passo (ficaria como pré-requisito) e o passo de `push` faria apenas a publicação do container, o que acha?
> 2. Nessas linhas ([aqui](https://github.com/mmmarceleza/iac-docker-tika/blob/babba72e1e8834415656f3db3b627e8715a72f9f/.github/workflows/main.yml#L78) e [aqui](https://github.com/mmmarceleza/iac-docker-tika/blob/babba72e1e8834415656f3db3b627e8715a72f9f/.github/workflows/main.yml#L83)) não seria melhor nomear com o nome certinho do repo? algo como `iac-docker-tika:....`
> 3. No seu repositório o CI está falhando, pode dar uma olhada nisso?

Fala @edsoncelio , tudo bom? Grato pela análise feita nas modificações. Vou acertar esses pontos amanhã, na dúvida te pingo. O item 3 foi o que mais me encabulou, cadastrei as secrets do meu Docker Hub pessoal e não funcionou. Amanhã faço novas tentativas. 

Valeu!

mmmarceleza committed:
atualizacao do nome da imagem

Por sugestão do @edsoncelio, atualizei o nome da imagem no scan.

mmmarceleza committed:
update tika version to 1.27

Com a versão 1.26 o scan detectou uma vulnerabilidade do nível HIGH.

mmmarceleza commented:
> Muito bom @mmmarceleza, seguem alguns pontos que identifiquei/fiquei em dúvida:
> 
> 1. Dado que o passo de `scan` vai executar sempre, poderíamos deixar a execução apenas em `pull_request`, com isso a gente ia garantir que `push` em `main` vai sempre ter executado esse passo (ficaria como pré-requisito) e o passo de `push` faria apenas a publicação do container, o que acha?
> 2. Nessas linhas ([aqui](https://github.com/mmmarceleza/iac-docker-tika/blob/babba72e1e8834415656f3db3b627e8715a72f9f/.github/workflows/main.yml#L78) e [aqui](https://github.com/mmmarceleza/iac-docker-tika/blob/babba72e1e8834415656f3db3b627e8715a72f9f/.github/workflows/main.yml#L83)) não seria melhor nomear com o nome certinho do repo? algo como `iac-docker-tika:....`
> 3. No seu repositório o CI está falhando, pode dar uma olhada nisso?

@edsoncelio , fiquei na dúvida em relação ao item 1, poderia dar mais detalhes? Quanto aos outros pontos eu já atendi, corrigi o CI e troquei o nome da imagem na etapa do scan. 

Curiosidade: a versão que não tinha vulnerabilidade dos níveis HIGH e CRITICAL, foi lançada a 4 dias atrás.

somatorio reviewed (approved):
O item levantado pelo @edsoncelio me pareceu mais uma melhoria do que requisito, então acho que podemos fazer o merge agora para já termos isso e fazer um novo PR depois com a mudança proposta
